Split up the Full Node and Validator functionality into two separate classes

TrustHenry opened the pull request:
We want to be a mixed network configuration of Fullnode and Validator.
The function of the Validator is not required when operating as a fullnode.
Each function can be separated to remove or add unnecessary functions.

~~Currently, this PR has two errors in tests~~
~~Error occurs in unittest in agora.test.Failures.~~
~~Also in the System Integration Test.~~
~~There seems to be a problem when a network test occurs.~~

Relates #586

Commit:
Split Node into FullNode and Validator

A FullNode should not contain any code related
to running the consensus protocol.

The separation of concerns avoid proliferation of
'if' statements in the Node class, and moves
the responsibility of the validator explicitly
into the Validator class.

In the unittests we instantiate RemoteAPI with
agora.API.Validator to avoid having to introduce
a lot of templated code and special cases.

When the NetworkManager tries to retrieve the PublicKey
of a node (which is actually a FullNode), the Node
in the unittests will return PublicKey.init.

In production, vibe.d will actually produce an
HTTPStatusException with status == 404,
so we handle it with a try/catch.

Also changed the system integration tests to
instantiate a FullNode (in addition to the existing
3 Validator nodes), to ensure the 404 HTTP status
handing code is actually executed.

Fixes #704

Co-Authored-By: Andrej Mitrovic <andrej.mitrovich@gmail.com>

AndrejMitrovic commented:
This is still work in progress. Do you want a review, or should we wait until it's finished?

TrustHenry commented:
> This is still work in progress. Do you want a review, or should we wait until it's finished?


I really want a review.  But there is a problem and I want to know what the problem is.

AndrejMitrovic commented on source/agora/node/FullNode.d:
```diff
@@ -181,7 +174,7 @@ public class Node : API
             bool isNominating ()
             {
-                return this.config.node.is_validator &&
+                return (this.nominator !is null) &&
```
A FullNode is not a Validator, right? Therefore it will never have a Nominator instance. So what is the purpose of this check here if it will never return true?

One option is to replace it with an inline delegate like so:

```D
// false: full node is never in the middle of nominating
this.network.startPeriodicCatchup(this.ledger, () => false);
```

But a better solution is to change `startPeriodicCatchup` to make the delegate optional.

AndrejMitrovic commented on source/agora/network/NetworkClient.d:
```diff
@@ -186,6 +187,9 @@ class NetworkClient
     {
         try
         {
+            if (!this.api.getNodeInfo().is_validator)
```
This change has several problems:
- sendEnvelope() was previously an asynchronous task. You have now turned it into a synchronous (blocking) task. Naturally this will break all the unittest code that uses timeouts.
- You should query the node whether it is a validator **once** after you connect to it. What is the purpose of asking it if it's still a validator every time we gossip an envelope? The node won't magically turn into a FullValidator while the connection is still active.

AndrejMitrovic commented on source/agora/network/NetworkManager.d:
```diff
@@ -375,7 +375,7 @@ public class NetworkManager
             try
             {
-                node.getPublicKey();
+                //node.getPublicKey();
```
This should be fixed.

AndrejMitrovic commented on source/agora/node/FullNode.d:
```diff
@@ -255,7 +242,7 @@ public class Node : API
 
             // then nominate
-            if (this.config.node.is_validator)
+            if (this.nominator !is null)
```
This belongs in the Validator class. The purpose of splitting the classes was to not have to do these checks.

https://github.com/bpfkorea/agora/issues/586

In the definition of done, the 4th item in the list is: 

> Fields which only belong in the Validator class should not be present in the FullNode class. For example, the Nominator instance should only be visible in the Validator class.

So remove the `nominator` instance from the FullNode class.

AndrejMitrovic commented on source/agora/node/FullNode.d:
```diff
@@ -572,7 +531,7 @@ public class Node : API
 *******************************************************************************/
 
-public Node runNode (Config config)
+public Node runNode (Node) (Config config)
```
Good change, but the documentation needs to be updated.

AndrejMitrovic commented on source/agora/node/Validator.d:
```diff
@@ -0,0 +1,95 @@
+    public this (const Config config)
+    {
+        super(config);
+        if (this.config.node.is_validator)
```
Why? This class **is** a validator. Remove these `if` checks.

AndrejMitrovic commented on source/agora/node/Validator.d:
```diff
@@ -0,0 +1,95 @@
+    public void receiveEnvelope (SCPEnvelope envelope)
+    {
+        // we should not receive SCP messages unless we're a validator node
+        if (!this.config.node.is_validator)
```
Remove this.

AndrejMitrovic commented on source/agora/node/Validator.d:
```diff
@@ -0,0 +1,95 @@
+        this.enroll_man.getValidatorPreimage(enroll_key, preimage);
+        return preimage;
+    }
+
```
```suggestion
```

AndrejMitrovic commented on source/agora/node/Validator.d:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    /// GET: /get_preimage
+    public PreimageInfo getPreimage (Hash enroll_key)
```
`override`

AndrejMitrovic commented on source/agora/node/Validator.d:
```diff
@@ -0,0 +1,95 @@
+
+    ***************************************************************************/
+
+    public void receiveEnvelope (SCPEnvelope envelope)
```
`override`

AndrejMitrovic commented on source/agora/node/Validator.d:
```diff
@@ -0,0 +1,95 @@
+    }
+
+    /// GET /public_key
+    public PublicKey getPublicKey () pure nothrow @safe @nogc
```
`override`

MichaelKim20 commented on source/agora/cli/multi/main.d:
```diff
@@ -13,9 +13,9 @@
 module agora.cli.multi.main;
 
-import agora.api.FullNode;
+import agora.api.Validator;
```
I think the following source needs to be fixed in order to run normally.

``` D
import agora.api.Validator;
...
import agora.node.FullNode;
import agora.node.Validator;
...
    Validator[] nodes;
    foreach (const ref config; configs)
        nodes ~= runNode!Validator(config);
...
}
```

MichaelKim20 reviewed (changes requested):
There is one modification in the CLI.

MichaelKim20 commented:
To compile agora-client, `dub.json` must modify it as follows.

```
        {
            "name": "client",
            "targetName": "agora-client",
            "mainSourceFile": "source/agora/cli/client/main.d",
            "excludedSourceFiles": [
                "source/agora/node/main.d",
                "source/agora/cli/multi/*"
            ]
        },
```

Geod24 commented on dub.json:
```diff
@@ -87,16 +87,7 @@
-                "source/agora/utils/SCPPrettyPrinter.d",
-                "source/scpp/*",
-                "source/scpd/*",
-                "source/agora/consensus/protocol/*",
```
That doesn't look right.

TrustHenry replied:
Re-add only that part to the ExcludedSourceFiles

Geod24 commented on source/agora/api/FullNode.d:
```diff
@@ -201,23 +204,4 @@ public interface API
-
-    ***************************************************************************/
-
-    public PreImageInfo getPreimage (Hash enroll_key);
```
This doesn't look right either. The full nodes should be able to propagate preimages too.

TrustHenry replied:
They're back in place.

TrustHenry replied:
This PR now spreads preimages, including full nodes.

Geod24 commented on source/agora/cli/multi/main.d:
```diff
@@ -13,9 +13,10 @@
+import agora.api.Validator;
 import agora.common.Config;
-import agora.node.Node;
+import agora.node.FullNode;
```
This shouldn't be needed, is it ?

TrustHenry replied:
Yes,

Geod24 commented on source/agora/cli/multi/main.d:
```diff
@@ -13,9 +13,10 @@
 module agora.cli.multi.main;
 
-import agora.api.FullNode;
+import agora.api.Validator;
```
I don't think the import is necessary, actually. Where is it used ?

Geod24 commented on source/agora/network/NetworkManager.d:
```diff
@@ -502,12 +502,11 @@ public class NetworkManager
+        pure nothrow @safe @nogc
     {
-        return NodeInfo(
-            this.minPeersConnected()
-                ? NetworkState.Complete : NetworkState.Incomplete,
-            this.known_addresses);
+        state = this.minPeersConnected() ? NetworkState.Complete : NetworkState.Incomplete;
+        addresses = this.known_addresses;
```
Looks unrelated (bad rebase?)

TrustHenry replied:
This is the first commit.
I excluded it because it is not relevant.

Geod24 commented on source/agora/node/FullNode.d:
```diff
@@ -568,11 +495,12 @@ public class Node : API
     The initialization process of the node is then completed.
 
     Params:
+      Node = FullNode, Validator Template is supported
```
```suggestion
      Node = Type of node to spawn (`FullNode` or `Validator`)
```

TrustHenry replied:
Thank you.

Geod24 commented on source/agora/node/FullNode.d:
```diff
@@ -541,21 +475,14 @@ public class Node : API
+    ///
+    protected void callTryNominate () @safe
     {
-        PreImageInfo preimage;
-        this.enroll_man.getValidatorPreimage(enroll_key, preimage);
-        return preimage;
     }
```
This `callTryNominate` looks suspicious.

TrustHenry replied:
Minor changes and names have been changed.

MichaelKim20 commented:
Please look at the source code below.
https://github.com/MukeunKim/agora-test/blob/master/source/app.d
Error occurs when invoking function of ValidatorAPI.

```
core.exception.AssertError@../../../../.dub/packages/vibe-core-1.9.1/vibe-core/source/vibe/core/log.d(895): Not Found (404)
```

Is it possible to use two API interfaces in vibe.d?

AndrejMitrovic commented:
> This doesn't look right either. The full nodes should be able to propagate preimages too.

Oh right, because they will need it to be able to validate blocks.

AndrejMitrovic commented:
> Please look at the source code below.

@MukeunKim your link is also a 404. :D

MichaelKim20 commented:
> > Please look at the source code below.
> 
> @MukeunKim your link is also a 404. :D

@AndrejMitrovic I changed it to the public.

AndrejMitrovic commented:
@MukeunKim: This fixes it:

```diff
diff --git a/source/agora/node/Node.d b/source/agora/node/Node.d
index c04f291a..eb6014c6 100644
--- a/source/agora/node/Node.d
+++ b/source/agora/node/Node.d
@@ -638,7 +638,7 @@ class ValidatorNode : FullNode, ValidatorAPI
 unittest
 {
     auto router2 = new URLRouter;
-    router2.registerRestInterface(new ValidatorNode);
+    router2.registerRestInterface!ValidatorAPI(new ValidatorNode);
     auto settings2 = new HTTPServerSettings;
     settings2.port = 8002;
     settings2.bindAddresses = ["::1", "127.0.0.1"];
```

I think the issue is that `registerRestInterface` doesn't look into the base class for all the APIs the class implements. Maybe this is a vibe.d bug. However, you can work around it with the change I showed here.

Geod24 commented:
> Is it possible to use two API interfaces in vibe.d?

Do you get a warning while building ? [This kind](https://github.com/vibe-d/vibe.d/issues/2414) ?

MichaelKim20 commented:
> > Is it possible to use two API interfaces in vibe.d?
> 
> Do you get a warning while building ? [This kind](https://github.com/vibe-d/vibe.d/issues/2414) ?

``` D
interface FullNodeAPI
{
    int getValue1();
}

interface ValidatorAPI : FullNodeAPI
{
    int getValue2();
}

class FullNode : FullNodeAPI
{
    public int getValue1()
    {
        return 1;
    }
}

class ValidatorNode : FullNode, ValidatorAPI
{
    public int getValue2()
    {
        return 2;
    }
}
```
Create a server using `ValidatorNode`.
HTTP404 error occurs when invoking the method `getValue2` that is only in interface `ValidatorAPI`.

AndrejMitrovic commented on source/agora/test/Failures.d:
```diff
@@ -30,6 +30,7 @@ unittest
 
     TestConf conf = { threshold : 100, timeout : 1_000 };
     auto network = makeTestNetwork(conf);
+    scope (exit) network.printLogs();
```
@TrustHenry we should remove this, because the nodes will be `null` after the constructor fails, so we cannot call their `printLog` method.

AndrejMitrovic commented:
This is currently WIP, don't review yet.

AndrejMitrovic commented:
So here's what we need to do:

- Change `getClient` to return a `ValidatorAPI` interface. This means we don't have to use templates in `Base.d`.
- Change `NetworkManager` to contain a map of PubKey => NetworkClient
- When we connect to each node we ask for its public key. If we get a 404 error it means this is a FullNode, not a validator. If we get a public key it means we can put the Key/ValidatorAPI into the map.
- Change `NetworkManager` to propagate validator messages to only validator nodes, and only to the ones in our quorum configuration. This means using the map that we just talked about. So for example, `gossipEnvelope` would only emit `SCPEnvelope` to nodes which are in that map

Personally I think we would need two maps:

```
Address[PubKey] validator_to_addr;  // validator to address
NetworkClient[Address] addr_to_client;  // address to NetworkClient
```

We would always add nodes to the `addr_to_client` map (because all nodes have an address), and we would add only Validator nodes to the `validator_to_addr` map. Then for each Validator in `validator_to_addr` you can find its NetworkClient in the `addr_to_client` map.

Maybe @Geod24 has a better idea than me though.

TrustHenry commented:
> When we connect to each node we ask for its public key. If we get a 404 error it means this is a FullNode, not a validator. If we get a public key it means we can put the Key/ValidatorAPI into the map.

If It request a public key when connecting to each node and distinguish the Validator by 404 error only,
Even if It are really a validator, you may be excluded from the 404 error code as a instant network sequence. Is it right to be excluded until the next request?

AndrejMitrovic commented:
404 means the node does not implement the API method. If there was just a network connection issue it would cause a request timeout and that's different.

Is this what you were asking?

TrustHenry commented:
And wait for the timeout until the error code for 404 appears.
What happens when this accumulates?

TrustHenry commented:
> 404 means the node does not implement the API method. If there was just a network connection issue it would cause a request timeout and that's different.
> 
> Is this what you were asking?

Yes, How about `403` or `405` `Method Not Allowed` with HTTP response code?

AndrejMitrovic commented:
> Yes, How about 403 or 405 Method Not Allowed with HTTP response code?

https://github.com/bpfkorea/vibe.d/blob/4076c01d438f39c358f976af69241243177c0d14/http/vibe/http/router.d#L27-L30

TrustHenry commented:
> > Yes, How about 403 or 405 Method Not Allowed with HTTP response code?
> 
> https://github.com/bpfkorea/vibe.d/blob/4076c01d438f39c358f976af69241243177c0d14/http/vibe/http/router.d#L27-L30

Oh, the default value is 404 which seems to be too common.

AndrejMitrovic commented:
You could do something like this:

```D
import vibe.http.common;

try
{
    pub_key = api.getPublicKey();
}
catch (HTTPStatusException ex)
{
    if (ex.status == 404)
    {
        // not a validator
    }
}
```

However, this will only work for system integration tests because they use vibe.d. When using LocalRest (in unittests) you will not get this exception thrown.

So we might have to change LocalRest. This code here: https://github.com/Geod24/localrest/blob/6da1455bb6793e76311acf4536be9506772e7372/source/geod24/LocalRest.d#L387

We could make it throw a custom exception type. For example we define `UnmatchedMethodException : Exception` in LocalRest, throw it, and then in Agora we could have:

```D
import vibe.http.common;

try
{
    pub_key = api.getPublicKey();
}
catch (HTTPStatusException ex)
{
    if (ex.status == 404)
    {
        // not a validator
    }
}
catch (UnhandledMethodException ex)
{
    // not a validator
}
```

Also seek input from @Geod24 because he's the LocalRest chief.

TrustHenry commented:
> We could make it throw a custom exception type. For example we define `UnmatchedMethodException : Exception` in LocalRest, throw it, and then in Agora we could have:
> 
> ```d
> import vibe.http.common;
> 
> try
> {
>     pub_key = api.getPublicKey();
> }
> catch (HTTPStatusException ex)
> {
>     if (ex.status == 404)
>     {
>         // not a validator
>     }
> }
> catch (UnhandledMethodException ex)
> {
>     // not a validator
> }
> ```
> 
> Also seek input from @Geod24 because he's the LocalRest chief.

@Geod24 ping

Geod24 commented:
Yes I think that makes sense.

TrustHenry commented:
Is there a solution for `agora.cli` link error?
Do I have to delete the `extludedSourceFiles` of the `dub.json` in `client`?

```
Linking...
Undefined symbols for architecture x86_64:
  "__D4scpd5types11Stellar_SCP12SCPQuorumSet11__xopEqualsFKxSQCdQCbQByQBoKxQpZb", referenced from:
      __D48TypeInfo_S4scpd5types11Stellar_SCP12SCPQuorumSet6__initZ in agora-client.o
  "__D4scpd5types11Stellar_SCP12SCPQuorumSet9__xtoHashFNbNeKxSQCeQCcQBzQBpZm", referenced from:
      __D48TypeInfo_S4scpd5types11Stellar_SCP12SCPQuorumSet6__initZ in agora-client.o
  "__D4scpd5types11Stellar_SCP12__ModuleInfoZ", referenced from:
      __D5agora3api9Validator12__ModuleInfoZ in agora-client.o
  "__D4scpd5types13Stellar_types9PublicKey11__xopEqualsFKxSQCbQBzQBwQBkKxQpZb", referenced from:
      __D46TypeInfo_S4scpd5types13Stellar_types9PublicKey6__initZ in agora-client.o
  "__D4scpd5types13Stellar_types9PublicKey9__xtoHashFNbNeKxSQCcQCaQBxQBlZm", referenced from:
      __D46TypeInfo_S4scpd5types13Stellar_types9PublicKey6__initZ in agora-client.o
  "__D5agora3api8FullNode12__ModuleInfoZ", referenced from:
      __D5agora6client13SendTxProcess12__ModuleInfoZ in agora-client.o
      __D5agora6client4main12__ModuleInfoZ in agora-client.o
      __D5agora3api9Validator12__ModuleInfoZ in agora-client.o
  "__D5agora3api8FullNode3API11__InterfaceZ", referenced from:
      __D5agora3api9Validator3API16__interfaceInfosZ in agora-client.o
      __D4vibe3web4rest__T19RestInterfaceClientTC5agora3api8FullNode3APIZQBv16__interfaceInfosZ in agora-client.o
  "void push_back<xdr::xvector<unsigned char, 4294967292u>, std::__1::vector<xdr::xvector<unsigned char, 4294967292u>, std::__1::allocator<xdr::xvector<unsigned char, 4294967292u> > > >(std::__1::vector<xdr::xvector<unsigned char, 4294967292u>, std::__1::allocator<xdr::xvector<unsigned char, 4294967292u> > >&, xdr::xvector<unsigned char, 4294967292u>&)", referenced from:
     
     __D4scpd3Cpp__T6vectorTSQv5types7XDRBase__T7xvectorThVki4294967292ZQyTSQCqQCo__T9allocatorTQCqZQpZQDf10fromStringFNfAyaZSQEoQEm__TQElTQEhTQCqZQEx in agora-client.o
  "void push_back<stellar::SCPQuorumSet, std::__1::vector<stellar::SCPQuorumSet, std::__1::allocator<stellar::SCPQuorumSet> > >(std::__1::vector<stellar::SCPQuorumSet, std::__1::allocator<stellar::SCPQuorumSet> >&, stellar::SCPQuorumSet&)", referenced from:
      __D4scpd3Cpp__T6vectorTSQv5types11Stellar_SCP12SCPQuorumSetTSQCgQCe__T9allocatorTQCgZQpZQCv10fromStringFNfAyaZSQEeQEc__TQEbTQDxTQCqZQEn in agora-client.o
  "void push_back<stellar::PublicKey, std::__1::vector<stellar::PublicKey, std::__1::allocator<stellar::PublicKey> > >(std::__1::vector<stellar::PublicKey, std::__1::allocator<stellar::PublicKey> >&, stellar::PublicKey&)", referenced from:
      __D4scpd3Cpp__T6vectorTSQv5types13Stellar_types9PublicKeyTSQCeQCc__T9allocatorTQCeZQpZQCt10fromStringFNfAyaZSQEcQEa__TQDzTQDvTQCqZQEl in agora-client.o
  "void push_back<unsigned char, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > >(std::__1::vector<unsigned char, std::__1::allocator<unsigned char> >&, unsigned char&)", referenced from:
      __D4scpd3Cpp__T6vectorThTSQxQu__T9allocatorThZQnZQBi10fromStringFNfAyaZSQCrQCp__TQCoThTQCkZQCy in agora-client.o
ld: symbol(s) not found for architecture x86_64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

TrustHenry commented:
> Is there a solution for `agora.cli` link error?
> Do I have to delete the `extludedSourceFiles` of the `dub.json` in `client`?

This problem has been solved.
I deleted it from ExcludedSourceFiles. "source/agora/api/FullNode.d"

Commit:
Move receivePreimage API to FullNode

FullNode should also receive preimages so it can validate blocks

MukeunKim commented:
> > Is there a solution for `agora.cli` link error?
> > Do I have to delete the `extludedSourceFiles` of the `dub.json` in `client`?
> 
> This problem has been solved.
> I deleted it from ExcludedSourceFiles. "source/agora/api/FullNode.d"

MichaelKim20 commented:
> Is there a solution for `agora.cli` link error?
> Do I have to delete the `extludedSourceFiles` of the `dub.json` in `client`?
> 
> ```
> Linking...
> Undefined symbols for architecture x86_64:
>   "__D4scpd5types11Stellar_SCP12SCPQuorumSet11__xopEqualsFKxSQCdQCbQByQBoKxQpZb", referenced from:
>       __D48TypeInfo_S4scpd5types11Stellar_SCP12SCPQuorumSet6__initZ in agora-client.o
>   "__D4scpd5types11Stellar_SCP12SCPQuorumSet9__xtoHashFNbNeKxSQCeQCcQBzQBpZm", referenced from:
>       __D48TypeInfo_S4scpd5types11Stellar_SCP12SCPQuorumSet6__initZ in agora-client.o
>   "__D4scpd5types11Stellar_SCP12__ModuleInfoZ", referenced from:
>       __D5agora3api9Validator12__ModuleInfoZ in agora-client.o
>   "__D4scpd5types13Stellar_types9PublicKey11__xopEqualsFKxSQCbQBzQBwQBkKxQpZb", referenced from:
>       __D46TypeInfo_S4scpd5types13Stellar_types9PublicKey6__initZ in agora-client.o
>   "__D4scpd5types13Stellar_types9PublicKey9__xtoHashFNbNeKxSQCcQCaQBxQBlZm", referenced from:
>       __D46TypeInfo_S4scpd5types13Stellar_types9PublicKey6__initZ in agora-client.o
>   "__D5agora3api8FullNode12__ModuleInfoZ", referenced from:
>       __D5agora6client13SendTxProcess12__ModuleInfoZ in agora-client.o
>       __D5agora6client4main12__ModuleInfoZ in agora-client.o
>       __D5agora3api9Validator12__ModuleInfoZ in agora-client.o
>   "__D5agora3api8FullNode3API11__InterfaceZ", referenced from:
>       __D5agora3api9Validator3API16__interfaceInfosZ in agora-client.o
>       __D4vibe3web4rest__T19RestInterfaceClientTC5agora3api8FullNode3APIZQBv16__interfaceInfosZ in agora-client.o
>   "void push_back<xdr::xvector<unsigned char, 4294967292u>, std::__1::vector<xdr::xvector<unsigned char, 4294967292u>, std::__1::allocator<xdr::xvector<unsigned char, 4294967292u> > > >(std::__1::vector<xdr::xvector<unsigned char, 4294967292u>, std::__1::allocator<xdr::xvector<unsigned char, 4294967292u> > >&, xdr::xvector<unsigned char, 4294967292u>&)", referenced from:
>      
>      __D4scpd3Cpp__T6vectorTSQv5types7XDRBase__T7xvectorThVki4294967292ZQyTSQCqQCo__T9allocatorTQCqZQpZQDf10fromStringFNfAyaZSQEoQEm__TQElTQEhTQCqZQEx in agora-client.o
>   "void push_back<stellar::SCPQuorumSet, std::__1::vector<stellar::SCPQuorumSet, std::__1::allocator<stellar::SCPQuorumSet> > >(std::__1::vector<stellar::SCPQuorumSet, std::__1::allocator<stellar::SCPQuorumSet> >&, stellar::SCPQuorumSet&)", referenced from:
>       __D4scpd3Cpp__T6vectorTSQv5types11Stellar_SCP12SCPQuorumSetTSQCgQCe__T9allocatorTQCgZQpZQCv10fromStringFNfAyaZSQEeQEc__TQEbTQDxTQCqZQEn in agora-client.o
>   "void push_back<stellar::PublicKey, std::__1::vector<stellar::PublicKey, std::__1::allocator<stellar::PublicKey> > >(std::__1::vector<stellar::PublicKey, std::__1::allocator<stellar::PublicKey> >&, stellar::PublicKey&)", referenced from:
>       __D4scpd3Cpp__T6vectorTSQv5types13Stellar_types9PublicKeyTSQCeQCc__T9allocatorTQCeZQpZQCt10fromStringFNfAyaZSQEcQEa__TQDzTQDvTQCqZQEl in agora-client.o
>   "void push_back<unsigned char, std::__1::vector<unsigned char, std::__1::allocator<unsigned char> > >(std::__1::vector<unsigned char, std::__1::allocator<unsigned char> >&, unsigned char&)", referenced from:
>       __D4scpd3Cpp__T6vectorThTSQxQu__T9allocatorThZQnZQBi10fromStringFNfAyaZSQCrQCp__TQCoThTQCkZQCy in agora-client.o
> ld: symbol(s) not found for architecture x86_64
> clang: error: linker command failed with exit code 1 (use -v to see invocation)
> ```

The cause of this error is when the SCP C++ library cannot be found.
SCP C++ library is required when referring to agora.api.Validator .

AndrejMitrovic committed:
Add missing @safe in 'receiveEnvelope'

AndrejMitrovic commented on source/agora/node/FullNode.d:
```diff
@@ -100,17 +98,25 @@ public class Node : API
+    ***************************************************************************/
+
+    public this (const Config config,
+        Set!PublicKey required_peer_keys = Set!PublicKey.init)
```
This is something I'd like to move out of the `FullNode` constructor, but I couldn't find a good solution yet.

AndrejMitrovic commented:
> node-3_1_8aa41cd509c7 | 2020-04-23T08:19:08.671319883Z Failed to parse config file 'config.yaml'. Error: Unable to open file config.yaml for YAML loading: config.yaml: No such file or directory

@TrustHenry the config file wasn't commited. I will push to your branch with the config file, ok?

AndrejMitrovic commented:
@TrustHenry you can `git fetch henry` or whatever your remote's name is. The file should be there now.

TrustHenry commented:
It's working now
@bpfkorea/core Ready for review.

AndrejMitrovic commented on source/agora/cli/multi/main.d:
```diff
@@ -104,9 +104,9 @@ private int main (string[] args)
+    Validator[] nodes;
     foreach (const ref config; configs)
-        nodes ~= runNode(config);
+        nodes ~= runNode!Validator(config);
```
I think this is wrong. The configs that are passed can be both Validators and FullNodes. So the code will have to check each config and spawn each type. accordingly.

MichaelKim20 replied:
The following files are also required when testing modifications.
ci/single_binary_test.d.
I think it needs to be corrected.

TrustHenry replied:
applied the code checks each config and spawn each type.

AndrejMitrovic commented on source/agora/network/NetworkManager.d:
```diff
@@ -230,8 +236,9 @@ public class NetworkManager
                     ledger.getBlockHeight() + 1,
                     blocks => blocks.all!(block =>
```
We should think about fixing this in the future. It does not short-circuit.

Also there will be many `isNominating is null` checks in a tight loop here.

AndrejMitrovic replied:
We can think about fixing this in the future. It's somewhat unrelated to this PR.

AndrejMitrovic commented on source/agora/node/Validator.d:
```diff
@@ -0,0 +1,164 @@
+
+        // build the list of required quorum peers to connect to
+        Set!PublicKey required_peer_keys;
+        void getNodes (in QuorumConfig conf, ref Set!PublicKey nodes)
```
I recommend moving this into a private static method called `buildRequiredKeys`. The constructor code is hard to follow when it has a big nested function like this one inside.

Edit: It's also ok to leave this in a future PR to fix it up.

TrustHenry replied:
Applied through additional commit.

AndrejMitrovic commented on source/agora/node/Validator.d:
```diff
@@ -0,0 +1,164 @@
+        Receive an SCP envelope.
+
+        API:
+            GET /envelope
```
Shouldn't this be `PUT`?

AndrejMitrovic commented on source/agora/node/main.d:
```diff
@@ -70,7 +72,10 @@ private int main (string[] args)
             return 0;
         }
-        runTask(() => node = runNode(config));
+        if (config.node.is_validator)
```
```suggestion

        if (config.node.is_validator)
```

AndrejMitrovic commented on source/agora/node/main.d:
```diff
@@ -61,7 +63,7 @@ private int main (string[] args)
     }
 
-    Node node;
+    FullNode node;
```
I would rename this to the `API` from `agora.api.FullNode`. Because technically it can be a Validator, and using the interface makes it more explicit that this is not necessarily always a `FullNode` object.

TrustHenry replied:
This seems to require sufficient discussion.

TrustHenry replied:
Currently, this the 66line refers to class `agora.node.FullNode`

AndrejMitrovic replied:
Ok it was just a suggestion, feel free to overrule. :)

AndrejMitrovic commented on source/agora/test/Base.d:
```diff
@@ -43,9 +41,14 @@ import agora.consensus.EnrollmentManager;
 import agora.network.NetworkManager;
 import agora.node.Ledger;
-import agora.node.Node;
+import agora.node.FullNode;
```
FullNode import should be moved before Ledger for alphabetic order.

AndrejMitrovic commented on source/agora/test/Base.d:
```diff
@@ -560,6 +571,64 @@ public class TestNode : Node, TestAPI
+        return PublicKey.init;
+    }
+
+    public override void receiveEnvelope(SCPEnvelope envelope) @safe
```
```suggestion
    public override void receiveEnvelope (SCPEnvelope envelope) @safe
```

And these last two methods need `///` comments.

AndrejMitrovic commented on source/agora/test/Base.d:
```diff
@@ -560,6 +571,64 @@ public class TestNode : Node, TestAPI
+/// A Validator which also implements test routines in TestAPI
+public class TestValidatorNode : Validator, TestAPI
+{
+    mixin TestNodeMixin!();
```
```suggestion
    ///
    mixin TestNodeMixin!();
```


This will expand the mixin in the autogenerated documentation.

AndrejMitrovic commented on source/agora/test/Base.d:
```diff
@@ -560,6 +571,64 @@ public class TestNode : Node, TestAPI
+/// A FullNode which also implements test routines in TestAPI
+public class TestFullNode : FullNode, TestAPI
+{
+    mixin TestNodeMixin!();
```
```suggestion
    ///
    mixin TestNodeMixin!();
```

MichaelKim20 commented on ci/system_integration_test.d:
```diff
@@ -28,7 +28,8 @@ immutable DockerComposeLogs = [ "docker-compose", "-f", ComposeFile, "logs", "-t
+                                    "http://127.0.0.1:4003"];
 immutable Cleanup = [ "rm", "-rf", IntegrationPath.buildPath("node/0/.cache/"),
                       IntegrationPath.buildPath("node/1/.cache/"),
                       IntegrationPath.buildPath("node/2/.cache/")];
```
```suggestion
                      IntegrationPath.buildPath("node/2/.cache/"),                                       
                      IntegrationPath.buildPath("node/3/.cache/")];
```

MichaelKim20 reviewed (changes requested):
Good. It needs one change.

Commit:
move and rename this into a private static method called

The constructor code is hard to follow when it has a big nested function like this one inside.

TrustHenry commented:
I've applied suggestions up to here.
@bpfkorea/core Ready for review.

AndrejMitrovic commented on source/agora/node/Validator.d:
```diff
@@ -0,0 +1,178 @@
+        Params:
+            conf = the node configuration
+            quorum_conf = the SCP quorum set configuration
+            nodes = Instance of Set!PublicKey The publickey of nodes
```
```suggestion
            nodes = will contain the set of public keys to connect to
```

TrustHenry replied:
apply the suggestion.

AndrejMitrovic commented:
Just two comments. And the PR will need a rebase due to conflicts.

Geod24 commented on tests/system/node/3/config.yaml:
```diff
@@ -0,0 +1,63 @@
+  retry_delay: 1.5
+  max_retries: 5
+  timeout: 1000
+  # This is a randomly generated keypair
+  # If this node is not a validator, this will be ignored
+  #
+  # DO NOT USE THOSE VALUES ANYWHERE
+  seed:    SAQEQ475K4BTKTW5RRZGX6NGRM37SOPIEM3OGVD2WNZMGZHIFDE65KQQ
```
This can be removed

Geod24 commented on tests/system/node/3/config.yaml:
```diff
@@ -0,0 +1,63 @@
+################################################################################
+admin:
+  enabled: false
+  address: 127.0.0.1
+  port:    2827
```
Remove

Geod24 commented on tests/system/node/3/config.yaml:
```diff
@@ -0,0 +1,63 @@
+  - http://node-1:3826
+  - http://node-2:4826
+
+################################################################################
+##                               Quorum slices                                ##
+################################################################################
+quorum:
+  threshold: 66%
+  nodes:
+    - GBUVRIIBMHKC4PE6BK7MO2O26U2NJLW4WGGWKLAVLAA2DLFZTBHHKOEK
+
```
That should be removed as well

AndrejMitrovic replied:
This will cause failures because the config parser always tries to parse quorum slices.

There is another issue about removing the quorum config from the config though.

TrustHenry replied:
If it remove this, Clearly, quorum configuration analysis error occurs.

Geod24 commented on source/agora/test/NetworkManager.d:
```diff
@@ -110,15 +110,24 @@ unittest
+
+            // the test has 3 nodes:
+            // 1 validator => used for creating blocks
+            // 1 byzantine FullNode => lies about the blockchain
```
```suggestion
            // 1 byzantine FullNode => lies about the blockchain (returns syntactically invalid data)
```

This is changing the meaning of this test a bit. Originally there were 2 validators and 1 byzantine node.
I don't mind though, but that might create issues with quorum sets.

AndrejMitrovic replied:
Actually I think the documentation is correct, it's just that it was undocumented before. Maybe this change should go into its own commit.

TrustHenry replied:
Add it as own commit.

Geod24 commented on source/agora/test/Base.d:
```diff
@@ -560,6 +571,68 @@ public class TestNode : Node, TestAPI
+        super(config);
+    }
+
+    /// FullNode does not implement this
+    public override void broadcastPreimage (uint height)
+    {
+        assert(0);
+    }
+
+    /// FullNode does not implement this
+    public override Enrollment createEnrollmentData ()
+    {
+        assert(0);
+    }
+
+    /// FullNode does not implement this
+    public override void updateEnrolledHeight (Hash enroll_key, ulong height)
+    {
+        assert(0);
+    }
```
Those methods shouldn't be available if they aren't implemented then

TrustHenry replied:
`public class TestFullNode : FullNode, TestAPI`
Here, `TestAPI` is inheritance interface `TestAPI: ValidatorAPI`
I'd like to delete that method.
But, I think there is nothing we can do about the current TestAPI inheritance structure.
Do you have a better idea?

Geod24 commented on source/agora/node/FullNode.d:
```diff
@@ -533,37 +450,15 @@ public class Node : API
+    protected void onAcceptedTransaction () @safe
+    {
+    }
 }
```
Moving this could be its own commit

TrustHenry replied:
The existing code use `this.nominator.tryNominate()` in `FullNode.putTransaction`
Because the Nominator has been moved from the corresponding refectoring, it must be handled by the same commit.

TrustHenry replied:
Oh, I understood the point.
I separated own commit.

Geod24 commented:
Looks pretty nice overall, just a few comments

Commit:
Add description in createNewNode of test NetworkManager

TrustHenry commented:
I applied Suggestion & replied
@Geod24 ping

TrustHenry commented:
Rebase with the latest commit.

TrustHenry commented:
Rebased with the latest commit.

Commit:
Reduce duplication of code and distinguish between Fullnode and Validator functions

This is a no-op for FullNode. A Validator node overrides this behavior.

TrustHenry commented:
@Geod24 Ready for review
